pkg/controllermanager/controller/utils/finalizers.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// RemoveGardenerFinalizer removes the gardner finalizer from the object.
+func RemoveGardenerFinalizer(ctx context.Context, c client.Client, obj kutil.Object) error {
+	if err := kutil.TryUpdate(ctx, retry.DefaultRetry, c, obj, func() error {
+		finalizers := sets.NewString(obj.GetFinalizers()...)
+		finalizers.Delete(gardenv1beta1.GardenerName)
+		obj.SetFinalizers(finalizers.UnsortedList())
+		return nil
+	}); err != nil {
+		return fmt.Errorf("could not remove \"%s\" finalizer: %+v", gardenv1beta1.GardenerName, err)
+	}
+
+	// Wait until the above modifications are reflected in the cache to prevent unwanted reconcile
+	// operations (sometimes the cache is not synced fast enough).
+	return wait.PollImmediate(time.Second, 30*time.Second, func() (bool, error) {
```

Comment:
Better use the context with a timeout, use `wait.Until`

Reply:
We want poller which exits when it succeeds and not the periodic job which will execute untill timeout. But yes, will change to pass parent context with timeout, to not miss the stop signal from parent context..

Reply:
Thanks!